CMakeLists.txt

```diff
@@ -639,6 +640,49 @@ PUBLIC
 
 add_subdirectory(interface/basisu_c_binding)
 
+# Only one architecture is supported at once, if neither of
+# ISA_SSE41 and ISA_SSE2  are defined ISA_AVX2 is chosen.
+# If ISA_AVX2 fails to compile user must chose other x86 options.
+# On arm based systems ISA_NEON is default
+
+list(FIND CMAKE_OSX_ARCHITECTURES "$(ARCHS_STANDARD)" ASTC_BUILD_UNIVERSAL)
+
+if(${ASTC_BUILD_UNIVERSAL} EQUAL -1)
+    if (${ISA_NONE})
+        set(ASTC_LIB_NAME astcenc-none-static)
+    else()
+        if(CPU_ARCHITECTURE STREQUAL x86_64 OR CPU_ARCHITECTURE STREQUAL x86)
+            if (${ISA_SSE41})
+                set(ASTC_LIB_NAME astcenc-sse4.1-static)
+            elseif (${ISA_SSE2})
+                set(ASTC_LIB_NAME astcenc-sse2-static)
+            else()
+                set(ISA_AVX2 ON)
+                set(ASTC_LIB_NAME astcenc-avx2-static)
+            endif()
+            if(CPU_ARCHITECTURE STREQUAL x86)
+                set(ISA_NONE ON)
+                set(ISA_AVX2 OFF)
+                set(ASTCENC_POPCNT 0)
+                set(ASTC_LIB_NAME astcenc-none-static)
+            endif()
+        elseif(CPU_ARCHITECTURE STREQUAL armv8 OR CPU_ARCHITECTURE STREQUAL arm64)
+            set(ASTC_LIB_NAME astcenc-neon-static)
+        else()
+            message(STATUS "Unsupported ISA for ${CPU_ARCHITECTURE} arch, using ISA_NONE.")
+            set(ISA_NONE ON)
+        endif()
+    endif()
+else()
+    set(ASTC_LIB_NAME astcenc-static)
```

Comment:
In this case - universal build - is the ISA determined from the compiler pre-defined macros?

Reply:
Not exactly. There is quite a bit of complex logic in [here](https://github.com/KhronosGroup/KTX-Software/blob/master/lib/astc-encoder/Source/cmake_core.cmake#L173) but in a nutshell universal build only means we are building for arm64 and x86_64 together. You can still enable your `ISA_NEON`, `ISA_SSE/AVX` options. For arm64 NEON is the default in universal build.

Reply:
Ahh! So it a universal build not a universal binary. Do the ARM Limited compilation tools support universal (fat) binaries?

> EDIT: This however reminds me. Should there be a ISA_NONE added to the non-SSE builds here?. At the moment it work because SSE4.1 is default for universal build for the rest of the build however BASIS is built without it and there is no conflict. Is that the intension?

I think non-SSE builds should not use SSE anywhere. Oh! The CMake variable name is BASISU_SUPPORT_SSE. The behavior you describe is likely what would be expected from that name.

For the macOS universal binary builds SSE must be off since we only get to pass 1 set of -D defines and options to the compiler which are used for both Intel and ARM builds. The universal binary (non-SSE) build is working in CI for this PR so you must be doing something already to disable SSE for astc-encoder. Enabling use of SIMD in universal binaries is the main driver for wanting compile-time detection of what SIMD ISA can be used. Please impress that upon the people responsible for the encoder. Other clients will want to support macOS universal binaries.

I want to change BASISU_SUPPORT_SSE and make the replacement apply to all encoders and architectures so something like KTX_SUPPORT_SIMD. Let's leave this PR as is for now and fix this in a later PR along with getting SIMD working in universal binaries.

Please mark this conversation as resolved once you've read this. I suspect if I do it now, all you will see the collapsed conversation with the "Show Resolved" button.